src/Database/LSMTree/Internal/WriteBufferWriter.hs

```diff
@@ -0,0 +1,236 @@
+  -> SerialisedKey
+  -> Entry SerialisedValue (RawBlobRef m h)
+  -> m ()
+addKeyOp WriteBufferWriter{..} key op = do
+  -- TODO: consider optimisation described in 'Database.LSMTree.Internal.RunBuilder.addKeyOp'.
+  op' <- traverse (copyBlob writerHasFS writerBlobOffset (forWriteBufferBlob writerHandles)) op
+  if PageAcc.entryWouldFitInPage key op
+    then do
+      mPage <- ST.stToIO $ addSmallKeyOp writerPageAcc key op'
+      for_ mPage $ writeRawPage writerHasFS (forWriteBufferKOps writerHandles)
+    else do
+      (pages, overflowPages) <- ST.stToIO $ addLargeKeyOp writerPageAcc key op'
+      -- TODO: consider optimisation described in 'Database.LSMTree.Internal.RunBuilder.addKeyOp'.
+      for_ pages $ writeRawPage writerHasFS (forWriteBufferKOps writerHandles)
+      writeRawOverflowPages writerHasFS (forWriteBufferKOps writerHandles) overflowPages
+
+-- | See 'Database.LSMTree.Internal.RunAcc.addSmallKeyOp'.
+addSmallKeyOp ::
+     PageAcc s
+  -> SerialisedKey
+  -> Entry SerialisedValue BlobSpan
+  -> ST s (Maybe RawPage)
+addSmallKeyOp pageAcc key op =
+  assert (PageAcc.entryWouldFitInPage key op) $ do
+    pageBoundaryNeeded <-
+        -- Try adding the key/op to the page accumulator to see if it fits. If
+        -- it does not fit, a page boundary is needed.
+        not <$> PageAcc.pageAccAddElem pageAcc key op
+    if pageBoundaryNeeded
+      then do
+        -- We need a page boundary. If the current page is empty then we have
+        -- a boundary already, otherwise we need to flush the current page.
+        mPage <- flushPageIfNonEmpty pageAcc
+        -- The current page is now empty, either because it was already empty
+        -- or because we just flushed it. Adding the new key/op to an empty
+        -- page must now succeed, because we know it fits in a page.
+        added <- PageAcc.pageAccAddElem pageAcc key op
+        assert added $ pure mPage
+      else do
+        pure Nothing
+
+-- | See 'Database.LSMTree.Internal.RunAcc.addLargeKeyOp'.
+addLargeKeyOp
+  :: PageAcc s
+  -> SerialisedKey
+  -> Entry SerialisedValue BlobSpan -- ^ the full value, not just a prefix
+  -> ST s ([RawPage], [RawOverflowPage])
+addLargeKeyOp pageAcc key op =
+  assert (not (PageAcc.entryWouldFitInPage key op)) $ do
+    -- If the existing page accumulator is non-empty, we flush it, since the
+    -- new large key/op will need more than one page to itself.
+    mPagePre <- flushPageIfNonEmpty pageAcc
+    -- Make the new page and overflow pages. Add the span of pages to the index.
+    let (page, overflowPages) = PageAcc.singletonPage key op
+    -- Combine the results with anything we flushed before
+    let !pages = selectPages mPagePre page
+    return (pages, overflowPages)
+
+-- | Internal helper. See 'Database.LSMTree.Internal.RunAcc.flushPageIfNonEmpty'.
+flushPageIfNonEmpty :: PageAcc s -> ST s (Maybe RawPage)
+flushPageIfNonEmpty pageAcc = do
+    keysCount <- PageAcc.keysCountPageAcc pageAcc
+    if keysCount > 0
```

Comment:
To keep this code close to its original in `RunAcc`, it would have been better not to rename identifiers, like changing `nkeys` to `keysCount`. It doesn't have to change now, but something to keep in mind going forward